src/libraries/System.Private.CoreLib/src/System/MulticastDelegate.cs

```diff
@@ -0,0 +1,57 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Runtime.Serialization;
+
+namespace System
+{
+#pragma warning disable CS0660, CS0661 // Defining operators, but not overriding Equals/GetHashCode
+    public abstract partial class MulticastDelegate : Delegate, ISerializable
+    {
+        // This constructor is called from the class generated by the
+        //    compiler generated code (This must match the constructor
+        //    in Delegate
+        [RequiresUnreferencedCode("The target method might be removed")]
+        protected MulticastDelegate(object target, string method) : base(target, method)
+        {
+        }
+
+        // This constructor is called from a class to generate a
+        // delegate based upon a static method name and the Type object
+        // for the class defining the method.
+        protected MulticastDelegate([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] Type target, string method) : base(target, method)
+        {
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool operator ==(MulticastDelegate? d1, MulticastDelegate? d2)
+        {
+            // Test d2 first to allow branch elimination when inlined for null checks (== null)
+            // so it can become a simple test
+            if (d2 is null)
+            {
+                return d1 is null;
+            }
+
+            return ReferenceEquals(d2, d1) ? true : d2.Equals((object?)d1);
```

Comment:
```suggestion
            if (ReferenceEquals(d1, d2))
            {
                return true;
            }

            return d2 is not null && d2.Equals((object?)d1);
```
Wouldn't this be better?

Reply:
"Test d2 first to allow branch elimination when inlined for null checks (== null) so it can become a simple test" comment above explains why the code is written the way it is written.

Reply:
> "Test d2 first to allow branch elimination when inlined for null checks (== null) so it can become a simple test" comment above explains why the code is written the way it is written.

`ReferenceEquals` let's it become a simple test too though with a constant null?

Reply:
Yes, this looks better. Would you like to submit a PR with this improvement?

This pattern came from https://github.com/dotnet/coreclr/pull/21765/ that applied this pattern in multiple places (the pattern was later modified in https://github.com/dotnet/coreclr/pull/21829 to avoid a potential breaking change). All these places can use the same improvement.